analysis/src/diff.rs

```diff
@@ -35,39 +53,102 @@ impl DiffAnalyzer {
     pub fn analyze_crate_source_diff(
         self,
         package: &PackageMetadata,
-    ) -> Result<Option<DiffReport>> {
+    ) -> Result<CrateSourceDiffReport> {
         let name = package.name().to_string();
         let version = package.version().to_string();
         let repository = match package.repository() {
-            Some(repo) => repo,
-            None => return Ok(None),
+            Some(repo) => Self::trim_remote_url(repo)?,
+            None => {
+                return Ok(CrateSourceDiffReport {
+                    name,
+                    version,
+                    ..Default::default()
+                });
+            }
         };
 
-        // TODO return defualt report if below operations fail?
-
         //Setup a git repository for crates.io hosted source code
         let crate_source_path = self.get_cratesio_version(&name, &version)?;
         let crate_repo = self.init_git(&crate_source_path)?;
         let crate_repo_head = crate_repo.head()?.peel_to_commit()?;
 
         // Get commit for the version release in the git source
         let git_repo = self.get_github_repo(&name, &repository)?;
-        let head_commit_oid = self
-            .get_head_commit_oid_for_version(&git_repo, &name, &version)?
-            .ok_or_else(|| anyhow!("cannot get head commit for version"))?;
+        let head_commit_oid =
+            match self.get_head_commit_oid_for_version(&git_repo, &name, &version)? {
+                Some(commit) => commit,
+                None => {
+                    return Ok(CrateSourceDiffReport {
+                        name,
+                        version,
+                        release_commit_inferred: Some(false),
+                        ..Default::default()
+                    });
+                }
+            };
 
         // Add git repo as a remote to crate repo
-        self.setup_remote(&crate_repo, repository)?;
+        self.setup_remote(&crate_repo, &repository[..])?;
 
         // At this point, crate_repo contains crate.io hosted source with a single commit
         //                and git source as a remote
         // Therefore, we can get diff between crate_repo master and any remote commit
-        let git_version_commit = crate_repo.find_commit(head_commit_oid)?;
-        self.analyze_diff(&crate_repo, &crate_repo_head, &git_version_commit)?;
 
-        Ok(Some(DiffReport {
-            ..Default::default()
-        }))
+        // Get release version commit within crate repo
+        let git_version_commit = match crate_repo.find_commit(head_commit_oid) {
+            Ok(commit) => commit,
+            Err(_error) => {
+                // It is possilbe that the commit object does not belong to any branch in the repository,
+                // e.g., syn 0.15.44
+                // TODO: can we get around this to still get the diff?
```

Comment:
Would you be able to provide a bit more context on this error you're seeing?

Reply:
https://github.com/dtolnay/syn/commit/6d798b63c255e90b7b1dbbfb3707fdce1704a18d - I checked on the local git as well that the commit does not belong to any branch. My guess is that the commit/branch was later squashed or deleted. Therefore, `repo.find_commit()` fails to find to commit given the sha.

Reply:
Interesting...One thing that we could try, before bailing out, would be to take the tag name and try to fetch the tag directly. Doing the equivalent of `git fetch <SHA | tag name>` since presumably you have the tag name here or have been able to determine the commit SHA.